File: extension/rateLimiter.js
/**
 * Enterprise API Rate Limiter using Token Bucket Algorithm.
 * Implements token buckets, burst handling, fair usage policy across client tiers,
 * custom limit overrides, and usage analytics tracking for CmdBar.
 *
 * @module extension/rateLimiter
 */

/**
 * TokenBucket represents a single token bucket instance for rate limiting.
 * Handles token refills based on elapsed time and burst consumption.
 *
 * @public
 */
export class TokenBucket {
  /**
   * Constructs a new TokenBucket instance.
   *
   * @param {object} [options={}] - Bucket configuration options.
   * @param {number} [options.capacity=60] - Maximum capacity (burst size).
   * @param {number} [options.refillRate=1.0] - Refill rate in tokens per second.
   * @param {number} [options.initialTokens=null] - Initial tokens (defaults to capacity).
   * @public
   */
  constructor(options = {}) {
    const { capacity = 60, refillRate = 1.0, initialTokens = null } = options;
    this.capacity = Math.max(1, Number(capacity) || 60);
    this.refillRate = Math.max(0.0001, Number(refillRate) || 1.0);
    this.tokens = initialTokens !== null && initialTokens !== undefined
      ? Math.min(this.capacity, Math.max(0, Number(initialTokens)))
      : this.capacity;
    this.lastRefill = Date.now();
  }

  /**
   * Internal method to calculate elapsed time and refill tokens.
   *
   * @param {number} [now=Date.now()] - Current timestamp in milliseconds.
   * @returns {number} Current token count after refill.
   * @private
   */
  _refill(now = Date.now()) {
    const elapsedSec = Math.max(0, (now - this.lastRefill) / 1000);
    if (elapsedSec > 0) {
      this.tokens = Math.min(this.capacity, this.tokens + elapsedSec * this.refillRate);
      this.lastRefill = now;
    }
    return this.tokens;
  }

  /**
   * Attempts to consume the specified number of tokens from the bucket.
   *
   * @param {number} [cost=1] - Number of tokens to consume.
   * @param {number} [now=Date.now()] - Current timestamp in milliseconds.
   * @returns {object} Result object containing `allowed`, `remaining`, `capacity`, `resetInMs`, and `retryAfterSec`.
   * @public
   */
  tryConsume(cost = 1, now = Date.now()) {
    this._refill(now);
    const reqCost = Math.max(1, Number(cost) || 1);

    if (this.tokens >= reqCost) {
      this.tokens -= reqCost;
      return {
        allowed: true,
        remaining: this.tokens,
        capacity: this.capacity,
        resetInMs: this.getTimeToRefill(reqCost, now),
        retryAfterSec: 0
      };
    } else {
      const needed = reqCost - this.tokens;
      const resetMs = Math.ceil((needed / this.refillRate) * 1000);
      const retrySec = Math.ceil(needed / this.refillRate);

      return {
        allowed: false,
        remaining: this.tokens,
        capacity: this.capacity,
        resetInMs: resetMs,
        retryAfterSec: retrySec
      };
    }
  }

  /**
   * Retrieves the current available tokens in the bucket after refilling.
   *
   * @param {number} [now=Date.now()] - Current timestamp in milliseconds.
   * @returns {number} Available tokens.
   * @public
   */
  getTokens(now = Date.now()) {
    return this._refill(now);
  }

  /**
   * Calculates time in milliseconds required to refill up to the needed token count.
   *
   * @param {number} [neededTokens=1] - Target number of tokens needed.
   * @param {number} [now=Date.now()] - Current timestamp in milliseconds.
   * @returns {number} Time to refill in milliseconds.
   * @public
   */
  getTimeToRefill(neededTokens = 1, now = Date.now()) {
    this._refill(now);
    const target = Math.max(1, Number(neededTokens) || 1);
    if (this.tokens >= target) {
      return 0;
    }
    const needed = target - this.tokens;
    return Math.ceil((needed / this.refillRate) * 1000);
  }

  /**
   * Resets the bucket to full capacity.
   *
   * @param {number} [now=Date.now()] - Current timestamp in milliseconds.
   * @public
   */
  reset(now = Date.now()) {
    this.tokens = this.capacity;
    this.lastRefill = now;
  }
}

/**
 * APIRateLimiter provides centralized multi-tenant rate limiting with token buckets,
 * fair usage isolation, custom limits per client, tier management, and analytics.
 *
 * @public
 */
export class APIRateLimiter {
  /**
   * Constructs an APIRateLimiter instance.
   *
   * @param {object} [options={}] - Configuration options.
   * @param {number} [options.defaultCapacity=60] - Default capacity for unclassified clients.
   * @param {number} [options.defaultRefillRate=1.0] - Default refill rate (tokens/sec).
   * @param {string} [options.defaultTier="pro"] - Default tier for unclassified clients.
   * @param {object} [options.tiers={}] - Custom tier definitions map.
   * @public
   */
  constructor(options = {}) {
    this.defaultCapacity = Math.max(1, Number(options.defaultCapacity) || 60);
    this.defaultRefillRate = Math.max(0.0001, Number(options.defaultRefillRate) || 1.0);
    this.defaultTier = options.defaultTier || "pro";

    this.tiers = {
      free: { capacity: 10, refillRate: 0.1667 }, // ~10 req/min
      pro: { capacity: 60, refillRate: 1.0 },      // ~60 req/min
      enterprise: { capacity: 300, refillRate: 5.0 }, // ~300 req/min
      ...(options.tiers || {})
    };

    /** @type {Map<string, TokenBucket>} */
    this.buckets = new Map();

    /** @type {Map<string, { capacity?: number, refillRate?: number, tier?: string }>} */
    this.clientLimits = new Map();

    /** @type {Map<string, string>} */
    this.clientTiers = new Map();

    // Analytics metrics
    this.metrics = {
      totalRequests: 0,
      allowedRequests: 0,
      throttledRequests: 0,
      clientMetrics: {}
    };
  }

  /**
   * Defines or updates a subscription tier.
   *
   * @param {string} tierName - Name of the tier (e.g. "free", "pro", "enterprise").
   * @param {object} config - Tier configuration `{ capacity, refillRate }`.
   * @public
   */
  setTier(tierName, config = {}) {
    if (!tierName || typeof tierName !== "string") return;
    this.tiers[tierName] = {
      capacity: Math.max(1, Number(config.capacity) || 60),
      refillRate: Math.max(0.0001, Number(config.refillRate) || 1.0)
    };
  }

  /**
   * Retrieves configuration for a given tier.
   *
   * @param {string} tierName - Name of the tier.
   * @returns {object} Tier configuration.
   * @public
   */
  getTier(tierName) {
    if (tierName && this.tiers[tierName]) {
      return { ...this.tiers[tierName] };
    }
    return {
      capacity: this.defaultCapacity,
      refillRate: this.defaultRefillRate
    };
  }

  /**
   * Sets custom rate limits or assigned tier for a specific client.
   *
   * @param {string} clientId - Unique identifier for client, user, or IP.
   * @param {object} limitOptions - `{ capacity, refillRate, tier }`.
   * @public
   */
  setClientLimit(clientId, limitOptions = {}) {
    if (!clientId) return;
    const cleanId = String(clientId).trim();
    const existing = this.clientLimits.get(cleanId) || {};
    const updated = { ...existing };

    if (limitOptions.capacity !== undefined && limitOptions.capacity !== null) {
      updated.capacity = Math.max(1, Number(limitOptions.capacity));
    }
    if (limitOptions.refillRate !== undefined && limitOptions.refillRate !== null) {
      updated.refillRate = Math.max(0.0001, Number(limitOptions.refillRate));
    }
    if (limitOptions.tier) {
      updated.tier = String(limitOptions.tier);
      this.clientTiers.set(cleanId, updated.tier);
    }

    this.clientLimits.set(cleanId, updated);

    // If bucket already exists and limits changed, recreate or adjust bucket
    const keyPrefix = `${cleanId}:`;
    for (const [key, bucket] of this.buckets.entries()) {
      if (key.startsWith(keyPrefix) || key === cleanId) {
        if (updated.capacity !== undefined) bucket.capacity = updated.capacity;
        if (updated.refillRate !== undefined) bucket.refillRate = updated.refillRate;
      }
    }
  }

  /**
   * Assigns a client to a specific tier.
   *
   * @param {string} clientId - Unique identifier for client.
   * @param {string} tierName - Tier name.
   * @public
   */
  setClientTier(clientId, tierName) {
    if (!clientId) return;
    const cleanId = String(clientId).trim();
    const cleanTier = String(tierName).trim();
    this.clientTiers.set(cleanId, cleanTier);

    const tierCfg = this.getTier(cleanTier);
    this.setClientLimit(cleanId, {
      capacity: tierCfg.capacity,
      refillRate: tierCfg.refillRate,
      tier: cleanTier
    });
  }

  /**
   * Resolves rate limit parameters for a client ID.
   *
   * @param {string} clientId - Client identifier.
   * @returns {object} Limit params `{ capacity, refillRate, tier }`.
   * @public
   */
  resolveClientParams(clientId) {
    const cleanId = clientId ? String(clientId).trim() : "default";
    const custom = this.clientLimits.get(cleanId) || {};
    const assignedTier = custom.tier || this.clientTiers.get(cleanId) || this.defaultTier;
    const tierCfg = this.getTier(assignedTier);

    return {
      capacity: custom.capacity !== undefined ? custom.capacity : tierCfg.capacity,
      refillRate: custom.refillRate !== undefined ? custom.refillRate : tierCfg.refillRate,
      tier: assignedTier
    };
  }

  /**
   * Gets or creates a TokenBucket instance for a client and route.
   *
   * @param {string} clientId - Client identifier.
   * @param {string} [route="default"] - API route or command category name.
   * @param {number} [now=Date.now()] - Timestamp in milliseconds.
   * @returns {TokenBucket} TokenBucket instance.
   * @public
   */
  getBucket(clientId, route = "default", now = Date.now()) {
    const cleanId = clientId ? String(clientId).trim() : "default";
    const cleanRoute = route ? String(route).trim() : "default";
    const bucketKey = `${cleanId}:${cleanRoute}`;

    if (!this.buckets.has(bucketKey)) {
      const params = this.resolveClientParams(cleanId);
      const bucket = new TokenBucket({
        capacity: params.capacity,
        refillRate: params.refillRate
      });
      bucket.lastRefill = now;
      this.buckets.set(bucketKey, bucket);
    }

    return this.buckets.get(bucketKey);
  }

  /**
   * Evaluates rate limit status for a client without consuming tokens (peek).
   *
   * @param {string} clientId - Client identifier.
   * @param {object} [options={}] - Options `{ route, cost, now }`.
   * @returns {object} Rate limit evaluation result.
   * @public
   */
  checkLimit(clientId, options = {}) {
    const cleanId = clientId ? String(clientId).trim() : "default";
    const route = options.route || "default";
    const cost = Math.max(1, Number(options.cost) || 1);
    const now = options.now || Date.now();

    const params = this.resolveClientParams(cleanId);
    const bucket = this.getBucket(cleanId, route, now);
    const currentTokens = bucket.getTokens(now);

    const allowed = currentTokens >= cost;
    const resetMs = bucket.getTimeToRefill(cost, now);
    const retrySec = allowed ? 0 : Math.ceil((cost - currentTokens) / bucket.refillRate);

    return {
      allowed,
      remaining: currentTokens,
      capacity: bucket.capacity,
      resetInMs: resetMs,
      retryAfterSec: retrySec,
      clientId: cleanId,
      route,
      tier: params.tier
    };
  }

  /**
   * Consumes tokens for a request from a given client and route.
   * Updates analytics and metrics upon consumption.
   *
   * @param {string} clientId - Client identifier.
   * @param {object} [options={}] - Options `{ route, cost, now }`.
   * @returns {object} Result object `{ allowed, remaining, capacity, resetInMs, retryAfterSec, clientId, route, tier }`.
   * @public
   */
  consume(clientId, options = {}) {
    const cleanId = clientId ? String(clientId).trim() : "default";
    const route = options.route || "default";
    const cost = Math.max(1, Number(options.cost) || 1);
    const now = options.now || Date.now();

    const params = this.resolveClientParams(cleanId);
    const bucket = this.getBucket(cleanId, route, now);
    const result = bucket.tryConsume(cost, now);

    // Update global analytics
    this.metrics.totalRequests++;
    if (result.allowed) {
      this.metrics.allowedRequests++;
    } else {
      this.metrics.throttledRequests++;
    }

    // Update client-specific analytics
    if (!this.metrics.clientMetrics[cleanId]) {
      this.metrics.clientMetrics[cleanId] = {
        totalRequests: 0,
        allowedRequests: 0,
        throttledRequests: 0,
        lastRequestTime: now,
        tier: params.tier,
        routes: {}
      };
    }

    const cm = this.metrics.clientMetrics[cleanId];
    cm.totalRequests++;
    cm.lastRequestTime = now;
    cm.tier = params.tier;

    if (result.allowed) {
      cm.allowedRequests++;
    } else {
      cm.throttledRequests++;
    }

    if (!cm.routes[route]) {
      cm.routes[route] = { total: 0, allowed: 0, throttled: 0 };
    }
    cm.routes[route].total++;
    if (result.allowed) cm.routes[route].allowed++;
    else cm.routes[route].throttled++;

    return {
      allowed: result.allowed,
      remaining: result.remaining,
      capacity: result.capacity,
      resetInMs: result.resetInMs,
      retryAfterSec: result.retryAfterSec,
      clientId: cleanId,
      route,
      tier: params.tier
    };
  }

  /**
   * Enforces fair usage policy for a request.
   * Alias for `consume` with explicit fair usage semantics.
   *
   * @param {string} clientId - Client identifier.
   * @param {string} [route="default"] - Route or feature identifier.
   * @param {object} [options={}] - Additional options `{ cost, now }`.
   * @returns {object} Result object.
   * @public
   */
  enforceFairUsage(clientId, route = "default", options = {}) {
    return this.consume(clientId, { route, ...options });
  }

  /**
   * Calculates usage percentage for a client's bucket (0% = full bucket, 100% = empty bucket).
   *
   * @param {string} clientId - Client identifier.
   * @param {string} [route="default"] - Route name.
   * @param {number} [now=Date.now()] - Timestamp in milliseconds.
   * @returns {number} Usage percentage (0.0 to 100.0).
   * @public
   */
  getUsagePercentage(clientId, route = "default", now = Date.now()) {
    const bucket = this.getBucket(clientId, route, now);
    const currentTokens = bucket.getTokens(now);
    const used = Math.max(0, bucket.capacity - currentTokens);
    return Math.min(100, (used / bucket.capacity) * 100);
  }

  /**
   * Retrieves analytics and metric statistics.
   *
   * @param {string} [clientId=null] - Optional client ID to filter metrics.
   * @returns {object} Analytics statistics summary object.
   * @public
   */
  getAnalytics(clientId = null) {
    if (clientId) {
      const cleanId = String(clientId).trim();
      const cm = this.metrics.clientMetrics[cleanId] || {
        totalRequests: 0,
        allowedRequests: 0,
        throttledRequests: 0,
        lastRequestTime: null,
        tier: this.resolveClientParams(cleanId).tier,
        routes: {}
      };
      const usagePct = this.getUsagePercentage(cleanId);
      const throttleRatePct = cm.totalRequests > 0
        ? (cm.throttledRequests / cm.totalRequests) * 100
        : 0;

      return {
        clientId: cleanId,
        ...cm,
        throttleRatePercentage: Number(throttleRatePct.toFixed(2)),
        usagePercentage: Number(usagePct.toFixed(2))
      };
    }

    const total = this.metrics.totalRequests;
    const throttled = this.metrics.throttledRequests;
    const throttleRatePct = total > 0 ? (throttled / total) * 100 : 0;
    const activeClients = Object.keys(this.metrics.clientMetrics);

    let peakUsageClient = null;
    let maxRequests = 0;
    for (const [id, cm] of Object.entries(this.metrics.clientMetrics)) {
      if (cm.totalRequests > maxRequests) {
        maxRequests = cm.totalRequests;
        peakUsageClient = id;
      }
    }

    const clientDetails = {};
    for (const [id, cm] of Object.entries(this.metrics.clientMetrics)) {
      const params = this.resolveClientParams(id);
      const bucket = this.getBucket(id);
      clientDetails[id] = {
        totalRequests: cm.totalRequests,
        allowedRequests: cm.allowedRequests,
        throttledRequests: cm.throttledRequests,
        tier: cm.tier,
        currentTokens: bucket.getTokens(),
        capacity: bucket.capacity,
        usagePercentage: Number(this.getUsagePercentage(id).toFixed(2))
      };
    }

    return {
      totalRequests: total,
      allowedRequests: this.metrics.allowedRequests,
      throttledRequests: throttled,
      throttleRatePercentage: Number(throttleRatePct.toFixed(2)),
      activeClientsCount: activeClients.length,
      peakUsageClient,
      clientDetails
    };
  }

  /**
   * Resets all analytics and metrics counters.
   *
   * @public
   */
  resetAnalytics() {
    this.metrics = {
      totalRequests: 0,
      allowedRequests: 0,
      throttledRequests: 0,
      clientMetrics: {}
    };
  }

  /**
   * Resets the token bucket for a specific client and route.
   *
   * @param {string} clientId - Client identifier.
   * @param {string} [route="default"] - Route name.
   * @public
   */
  resetClientBucket(clientId, route = "default") {
    const cleanId = clientId ? String(clientId).trim() : "default";
    const cleanRoute = route ? String(route).trim() : "default";
    const bucketKey = `${cleanId}:${cleanRoute}`;
    if (this.buckets.has(bucketKey)) {
      this.buckets.get(bucketKey).reset();
    }
  }

  /**
   * Clears all token buckets and client limits.
   *
   * @public
   */
  clearAllBuckets() {
    this.buckets.clear();
    this.clientLimits.clear();
    this.clientTiers.clear();
    this.resetAnalytics();
  }
}

/**
 * Creates a default singleton instance of APIRateLimiter.
 *
 * @type {APIRateLimiter}
 * @public
 */
export const defaultRateLimiter = new APIRateLimiter();

File: companion/rate_limiter.py
#!/usr/bin/env python3
"""
Enterprise API Rate Limiting Module for CmdBar Python Companion.
Implements Token Bucket algorithm, burst handling, fair usage isolation,
custom limits, tier management, analytics, and middleware integration wrappers.
"""

import time
import math
import functools
from typing import Dict, Any, Optional, Callable, Tuple


class TokenBucket:
    """
    TokenBucket class representing a single rate limiting bucket.
    """
    def __init__(
        self,
        capacity: int = 60,
        refill_rate: float = 1.0,
        refill_interval_ms: float = 1000.0,
        initial_tokens: Optional[float] = None,
        time_provider: Optional[Callable[[], float]] = None
    ):
        self.capacity = max(1, capacity)
        self.refill_rate = max(0.0, float(refill_rate))
        self.refill_interval_ms = max(1.0, float(refill_interval_ms))
        self.time_provider = time_provider if time_provider is not None else time.time

        if initial_tokens is not None:
            self.tokens = min(float(self.capacity), max(0.0, float(initial_tokens)))
        else:
            self.tokens = float(self.capacity)

        self.last_refill_time = self.time_provider()

    def refill(self) -> None:
        """
        Refills tokens based on elapsed time since last refill.
        """
        now = self.time_provider()
        elapsed_sec = max(0.0, now - self.last_refill_time)
        elapsed_ms = elapsed_sec * 1000.0

        if elapsed_ms > 0:
            tokens_to_add = (elapsed_ms / self.refill_interval_ms) * self.refill_rate
            self.tokens = min(float(self.capacity), self.tokens + tokens_to_add)
            self.last_refill_time = now

    def try_consume(self, tokens_to_consume: int = 1) -> Dict[str, Any]:
        """
        Attempts to consume tokens from the bucket.
        """
        self.refill()
        amount = max(1, int(tokens_to_consume))

        if self.tokens >= amount:
            self.tokens -= amount
            missing_for_full = self.capacity - self.tokens
            reset_ms = (
                math.ceil((missing_for_full / self.refill_rate) * self.refill_interval_ms)
                if self.refill_rate > 0 else 0
            )
            return {
                "allowed": True,
                "remaining": math.floor(self.tokens),
                "limit": self.capacity,
                "reset_ms": reset_ms,
                "retry_after_ms": 0,
                "tokens_consumed": amount,
            }
        else:
            missing_tokens = amount - self.tokens
            retry_after_ms = (
                math.ceil((missing_tokens / self.refill_rate) * self.refill_interval_ms)
                if self.refill_rate > 0 else 0
            )
            missing_for_full = self.capacity - self.tokens
            reset_ms = (
                math.ceil((missing_for_full / self.refill_rate) * self.refill_interval_ms)
                if self.refill_rate > 0 else 0
            )
            return {
                "allowed": False,
                "remaining": math.floor(self.tokens),
                "limit": self.capacity,
                "reset_ms": reset_ms,
                "retry_after_ms": retry_after_ms,
                "tokens_consumed": 0,
            }

    def get_tokens(self) -> int:
        """
        Returns current available token count after refill.
        """
        self.refill()
        return math.floor(self.tokens)

    def reset(self) -> None:
        """
        Resets the bucket to maximum capacity.
        """
        self.tokens = float(self.capacity)
        self.last_refill_time = self.time_provider()


class ApiRateLimiter:
    """
    Enterprise API Rate Limiter Manager for Python.
    Manages token buckets per client, tier, and endpoint with custom rules and analytics.
    """
    def __init__(
        self,
        default_limit: Optional[Dict[str, Any]] = None,
        tiers: Optional[Dict[str, Dict[str, Any]]] = None,
        custom_limits: Optional[Dict[str, Dict[str, Any]]] = None,
        time_provider: Optional[Callable[[], float]] = None,
        enable_analytics: bool = True
    ):
        self.time_provider = time_provider if time_provider is not None else time.time
        self.enable_analytics = enable_analytics

        self.default_limit = {
            "capacity": 60,
            "refill_rate": 1.0,
            "refill_interval_ms": 1000.0,
        }
        if default_limit:
            self.default_limit.update(default_limit)

        self.tiers = {
            "enterprise": {"capacity": 1000, "refill_rate": 50.0, "refill_interval_ms": 1000.0},
            "pro": {"capacity": 300, "refill_rate": 10.0, "refill_interval_ms": 1000.0},
            "free": {"capacity": 60, "refill_rate": 1.0, "refill_interval_ms": 1000.0},
        }
        if tiers:
            self.tiers.update(tiers)

        self.custom_limits: Dict[str, Dict[str, Any]] = {}
        if custom_limits:
            self.custom_limits.update(custom_limits)

        self.buckets: Dict[str, TokenBucket] = {}

        self.analytics = {
            "total_requests": 0,
            "allowed_requests": 0,
            "throttled_requests": 0,
            "tokens_consumed": 0,
            "client_stats": {},
            "endpoint_stats": {},
            "start_time": self.time_provider(),
        }

    def set_custom_limit(self, key: str, limit_config: Dict[str, Any]) -> None:
        """
        Registers or updates a custom rate limit rule.
        """
        if key and isinstance(key, str):
            self.custom_limits[key] = limit_config

    def remove_custom_limit(self, key: str) -> bool:
        """
        Removes a custom rate limit rule.
        """
        if key in self.custom_limits:
            del self.custom_limits[key]
            return True
        return False

    def get_limit_config(self, key: Optional[str] = None, tier: Optional[str] = None) -> Dict[str, Any]:
        """
        Resolves the limit configuration for a key or tier.
        """
        if key and key in self.custom_limits:
            return self.custom_limits[key]
        if tier and tier in self.tiers:
            return self.tiers[tier]
        return self.default_limit

    def get_bucket(self, bucket_key: str, limit_config: Dict[str, Any]) -> TokenBucket:
        """
        Retrieves or creates a TokenBucket for a bucket key.
        """
        bucket = self.buckets.get(bucket_key)
        cap = limit_config.get("capacity", 60)
        rate = limit_config.get("refill_rate", 1.0)
        interval = limit_config.get("refill_interval_ms", 1000.0)

        if not bucket:
            bucket = TokenBucket(
                capacity=cap,
                refill_rate=rate,
                refill_interval_ms=interval,
                time_provider=self.time_provider,
            )
            self.buckets[bucket_key] = bucket
        elif (
            bucket.capacity != cap
            or bucket.refill_rate != rate
            or bucket.refill_interval_ms != interval
        ):
            bucket.capacity = cap
            bucket.refill_rate = rate
            bucket.refill_interval_ms = interval

        return bucket

    def check_rate_limit(
        self,
        client_id: str = "global",
        endpoint: str = "default",
        tokens_to_consume: int = 1,
        tier: Optional[str] = None,
        cost: Optional[int] = None,
    ) -> Dict[str, Any]:
        """
        Checks rate limit for a client and endpoint request.
        """
        actual_tokens = cost if cost is not None else tokens_to_consume
        bucket_key = f"{client_id}:{endpoint}" if endpoint != "default" else client_id
        limit_config = self.get_limit_config(endpoint, tier) or self.get_limit_config(client_id, tier)

        bucket = self.get_bucket(bucket_key, limit_config)
        result = bucket.try_consume(actual_tokens)

        headers = {
            "X-RateLimit-Limit": str(result["limit"]),
            "X-RateLimit-Remaining": str(result["remaining"]),
            "X-RateLimit-Reset": str(math.ceil(result["reset_ms"] / 1000.0)),
        }
        if not result["allowed"]:
            headers["Retry-After"] = str(math.ceil(result["retry_after_ms"] / 1000.0))

        if self.enable_analytics:
            self._record_analytics(client_id, endpoint, result, actual_tokens)

        return {
            "allowed": result["allowed"],
            "limit": result["limit"],
            "remaining": result["remaining"],
            "reset_ms": result["reset_ms"],
            "retry_after_ms": result["retry_after_ms"],
            "tokens_consumed": result["tokens_consumed"],
            "headers": headers,
            "client": client_id,
            "endpoint": endpoint,
            "bucket_key": bucket_key,
        }

    def _record_analytics(
        self,
        client_id: str,
        endpoint: str,
        result: Dict[str, Any],
        tokens_requested: int
    ) -> None:
        """
        Records analytics statistics for requests.
        """
        self.analytics["total_requests"] += 1
        if result["allowed"]:
            self.analytics["allowed_requests"] += 1
            self.analytics["tokens_consumed"] += result["tokens_consumed"]
        else:
            self.analytics["throttled_requests"] += 1

        c_stats = self.analytics["client_stats"].setdefault(
            client_id, {"total": 0, "allowed": 0, "throttled": 0, "tokens_consumed": 0, "last_request": 0}
        )
        c_stats["total"] += 1
        c_stats["last_request"] = self.time_provider()
        if result["allowed"]:
            c_stats["allowed"] += 1
            c_stats["tokens_consumed"] += result["tokens_consumed"]
        else:
            c_stats["throttled"] += 1

        e_stats = self.analytics["endpoint_stats"].setdefault(
            endpoint, {"total": 0, "allowed": 0, "throttled": 0, "tokens_consumed": 0, "last_request": 0}
        )
        e_stats["total"] += 1
        e_stats["last_request"] = self.time_provider()
        if result["allowed"]:
            e_stats["allowed"] += 1
            e_stats["tokens_consumed"] += result["tokens_consumed"]
        else:
            e_stats["throttled"] += 1

    def get_quota_status(
        self, client_id: str = "global", endpoint: Optional[str] = None, tier: str = "free"
    ) -> Dict[str, Any]:
        """
        Inspects current quota status without consuming tokens.
        """
        ep_key = endpoint or "default"
        bucket_key = f"{client_id}:{ep_key}" if ep_key != "default" else client_id
        limit_config = self.get_limit_config(ep_key, tier) or self.get_limit_config(client_id, tier)

        bucket = self.get_bucket(bucket_key, limit_config)
        tokens = bucket.get_tokens()
        missing_for_full = bucket.capacity - tokens
        reset_ms = (
            math.ceil((missing_for_full / bucket.refill_rate) * bucket.refill_interval_ms)
            if bucket.refill_rate > 0 else 0
        )

        return {
            "client": client_id,
            "endpoint": ep_key,
            "tokens": tokens,
            "capacity": bucket.capacity,
            "refill_rate": bucket.refill_rate,
            "reset_ms": reset_ms,
            "is_throttled": tokens < 1,
        }

    def get_analytics(self) -> Dict[str, Any]:
        """
        Returns summary of rate limit analytics.
        """
        now = self.time_provider()
        uptime = now - self.analytics["start_time"]
        return {
            "total_requests": self.analytics["total_requests"],
            "allowed_requests": self.analytics["allowed_requests"],
            "throttled_requests": self.analytics["throttled_requests"],
            "tokens_consumed": self.analytics["tokens_consumed"],
            "active_buckets": len(self.buckets),
            "uptime_sec": uptime,
            "client_stats": {k: dict(v) for k, v in self.analytics["client_stats"].items()},
            "endpoint_stats": {k: dict(v) for k, v in self.analytics["endpoint_stats"].items()},
        }

    def get_client_stats(self, client_id: str) -> Optional[Dict[str, Any]]:
        """
        Returns stats for a specific client.
        """
        if client_id in self.analytics["client_stats"]:
            return dict(self.analytics["client_stats"][client_id])
        return None

    def reset_analytics(self) -> None:
        """
        Resets recorded analytics statistics.
        """
        self.analytics["total_requests"] = 0
        self.analytics["allowed_requests"] = 0
        self.analytics["throttled_requests"] = 0
        self.analytics["tokens_consumed"] = 0
        self.analytics["client_stats"].clear()
        self.analytics["endpoint_stats"].clear()
        self.analytics["start_time"] = self.time_provider()

    def reset_buckets(self) -> None:
        """
        Resets all managed token buckets.
        """
        for bucket in self.buckets.values():
            bucket.reset()

    def wrap_dbus_method(
        self,
        method_name: str,
        handler: Callable,
        client_id: str = "dbus-client",
        tier: Optional[str] = None,
        tokens_to_consume: int = 1,
        reject_with_error: bool = False,
    ) -> Callable:
        """
        Decorator/wrapper to apply rate limiting to D-Bus methods.
        """
        @functools.wraps(handler)
        def wrapper(*args, **kwargs):
            res = self.check_rate_limit(
                client_id=client_id,
                endpoint=method_name,
                tokens_to_consume=tokens_to_consume,
                tier=tier,
            )
            if not res["allowed"]:
                if reject_with_error:
                    raise RuntimeError(
                        f"Rate limit exceeded for {method_name}. Retry after {res['retry_after_ms']}ms."
                    )
                return False
            return handler(*args, **kwargs)

        return wrapper

    def wrap_ai_service(
        self,
        translate_fn: Callable,
        client_id: str = "ai-user",
        tier: str = "pro",
        tokens_to_consume: int = 5,
    ) -> Callable:
        """
        Decorator/wrapper for AI service calls with rate limiting.
        """
        @functools.wraps(translate_fn)
        def wrapper(*args, **kwargs):
            res = self.check_rate_limit(
                client_id=client_id,
                endpoint="ai-translate",
                tokens_to_consume=tokens_to_consume,
                tier=tier,
            )
            if not res["allowed"]:
                retry_sec = math.ceil(res["retry_after_ms"] / 1000.0)
                raise RuntimeError(f"AI rate limit exceeded. Retry after {retry_sec} seconds.")
            return translate_fn(*args, **kwargs)

        return wrapper
